src/ips.rs

```diff
@@ -0,0 +1,525 @@
+    }
+
+    fn degree_centrality_avg(&self, degrees: &HashMap<IpAddr, u32>) -> f64 {
+        (degrees.iter().fold(0, |acc, (_, &degree)| acc + degree) as f64) / degrees.len() as f64
+    }
+
+    fn rate_node(&self, node: &Node, degree: u32, eigenvalue: f64) -> f64 {
+        // Calculate rating for node (if min == max for normalization factors then rating is
+        // not increased for that factor as lerp() returns 0.0).
+        // Rating is a combination of the following factors:
+        let mut rating = 0.0;
+
+        // 1. Degree
+        rating += self.degree_factors.lerp(degree as f64)
+            * NORMALIZE_TO_VALUE
+            * self.config.mcda_weights.degree;
+
+        // 2. Betweenness
+        rating += self.betweenness_factors.lerp(node.betweenness)
+            * NORMALIZE_TO_VALUE
+            * self.config.mcda_weights.betweenness;
+
+        // 3. Closeness
+        rating += self.closeness_factors.lerp(node.closeness)
+            * NORMALIZE_TO_VALUE
+            * self.config.mcda_weights.closeness;
+
+        // 4. Eigenvector
+        rating += self.eigenvector_factors.lerp(eigenvalue)
+            * NORMALIZE_TO_VALUE
+            * self.config.mcda_weights.eigenvector;
+
+        rating
+    }
+
+    // Very simple algorithm to detect islands.
+    // Take first vertex and do BFS to find all connected vertices. If there are any unvisited vertices
+    // create new island and do BFS one more time. Repeat until all vertices are visited.
+    fn detect_islands(&self, agraph: &AGraph) -> Vec<HashSet<usize>> {
+        let mut islands = Vec::new();
+        let mut visited = vec![false; agraph.len()];
+
+        for i in 0..agraph.len() {
+            if visited[i] {
+                continue;
+            }
+
+            let mut island = HashSet::new();
+            let mut queue = VecDeque::new();
+            queue.push_back(i);
+
+            while let Some(node_idx) = queue.pop_front() {
+                if visited[node_idx] {
+                    continue;
+                }
+
+                island.insert(node_idx);
+
+                visited[node_idx] = true;
+
+                for j in 0..agraph[node_idx].len() {
+                    if !visited[agraph[node_idx][j]] {
+                        queue.push_back(agraph[node_idx][j]);
+                    }
+                }
+            }
+            islands.push(island);
+        }
+        islands
+    }
+}
+
+impl NormalizationFactors {
+    fn determine<T>(list: &[T]) -> NormalizationFactors
+    where
+        T: PartialOrd + Into<f64> + Copy,
+    {
+        let min = list
+            .iter()
+            .min_by(|a, b| a.partial_cmp(b).unwrap())
+            .unwrap();
+        let max = list
+            .iter()
+            .max_by(|a, b| a.partial_cmp(b).unwrap())
+            .unwrap();
+
+        NormalizationFactors {
+            min: (*min).into(),
+            max: (*max).into(),
+        }
+    }
+
+    fn lerp(&self, value: f64) -> f64 {
```

Comment:
Is this linear interpolation? Perhaps some doc would be helpful here as this isn't what is commonly referred to as `lerp` iirc?

Reply:
You're right, changed to a name that would fit better.